From fd483f831ff933e6303ebb9fb526232c5b9777d1 Mon Sep 17 00:00:00 2001
From: Martin Siggel <martin.siggel@dlr.de>
Date: Sat, 29 May 2021 01:50:17 +0200
Subject: [PATCH] Added dlr patch to support c2 continous coons patches

---
 src/SWIG_files/wrapper/GeomFill.i | 17 ++++++++++++-----
 1 file changed, 12 insertions(+), 5 deletions(-)

diff --git a/src/SWIG_files/wrapper/GeomFill.i b/src/SWIG_files/wrapper/GeomFill.i
index e4e38c43..ac0bafb4 100644
--- a/src/SWIG_files/wrapper/GeomFill.i
+++ b/src/SWIG_files/wrapper/GeomFill.i
@@ -108,6 +108,7 @@ enum GeomFill_FillingStyle {
 	GeomFill_StretchStyle = 0,
 	GeomFill_CoonsStyle = 1,
 	GeomFill_CurvedStyle = 2,
+	GeomFill_CoonsC2Style = 3,
 };
 
 enum GeomFill_PipeError {
@@ -1528,12 +1529,14 @@ C2: Geom_BSplineCurve
 C3: Geom_BSplineCurve
 C4: Geom_BSplineCurve
 Type: GeomFill_FillingStyle
+Tol: Standard_Real
+SetSameDistr: Standard_Boolean
 
 Returns
 -------
 None
 ") Init;
-		void Init(const opencascade::handle<Geom_BSplineCurve> & C1, const opencascade::handle<Geom_BSplineCurve> & C2, const opencascade::handle<Geom_BSplineCurve> & C3, const opencascade::handle<Geom_BSplineCurve> & C4, const GeomFill_FillingStyle Type);
+		void Init(const opencascade::handle<Geom_BSplineCurve> & C1, const opencascade::handle<Geom_BSplineCurve> & C2, const opencascade::handle<Geom_BSplineCurve> & C3, const opencascade::handle<Geom_BSplineCurve> & C4, const GeomFill_FillingStyle Type,  const Standard_Real Tol, const Standard_Boolean SetSameDistr);
 
 		/****************** Init ******************/
 		/**** md5 signature: 8c1e4ea27db140918013a210804afd7a ****/
@@ -6817,12 +6820,13 @@ P1: TColgp_Array1OfPnt
 P2: TColgp_Array1OfPnt
 P3: TColgp_Array1OfPnt
 P4: TColgp_Array1OfPnt
+isC2: Standard_Boolean
 
 Returns
 -------
 None
 ") GeomFill_Coons;
-		 GeomFill_Coons(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4);
+		 GeomFill_Coons(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const Standard_Boolean isC2 = Standard_False);
 
 		/****************** GeomFill_Coons ******************/
 		/**** md5 signature: cbef7bb663e9994379c9aab840ba0fd0 ****/
@@ -6839,12 +6843,13 @@ W1: TColStd_Array1OfReal
 W2: TColStd_Array1OfReal
 W3: TColStd_Array1OfReal
 W4: TColStd_Array1OfReal
+isC2: Standard_Boolean
 
 Returns
 -------
 None
 ") GeomFill_Coons;
-		 GeomFill_Coons(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const TColStd_Array1OfReal & W1, const TColStd_Array1OfReal & W2, const TColStd_Array1OfReal & W3, const TColStd_Array1OfReal & W4);
+		 GeomFill_Coons(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const TColStd_Array1OfReal & W1, const TColStd_Array1OfReal & W2, const TColStd_Array1OfReal & W3, const TColStd_Array1OfReal & W4, const Standard_Boolean isC2 = Standard_False);
 
 		/****************** Init ******************/
 		/**** md5 signature: d3ffe113a485ff934a537831bed2560b ****/
@@ -6857,12 +6862,13 @@ P1: TColgp_Array1OfPnt
 P2: TColgp_Array1OfPnt
 P3: TColgp_Array1OfPnt
 P4: TColgp_Array1OfPnt
+isC2: Standard_Boolean
 
 Returns
 -------
 None
 ") Init;
-		void Init(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4);
+		void Init(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const Standard_Boolean isC2 = Standard_False);
 
 		/****************** Init ******************/
 		/**** md5 signature: 094f730897c039878a8e655c32e65678 ****/
@@ -6879,12 +6885,13 @@ W1: TColStd_Array1OfReal
 W2: TColStd_Array1OfReal
 W3: TColStd_Array1OfReal
 W4: TColStd_Array1OfReal
+isC2: Standard_Boolean
 
 Returns
 -------
 None
 ") Init;
-		void Init(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const TColStd_Array1OfReal & W1, const TColStd_Array1OfReal & W2, const TColStd_Array1OfReal & W3, const TColStd_Array1OfReal & W4);
+		void Init(const TColgp_Array1OfPnt & P1, const TColgp_Array1OfPnt & P2, const TColgp_Array1OfPnt & P3, const TColgp_Array1OfPnt & P4, const TColStd_Array1OfReal & W1, const TColStd_Array1OfReal & W2, const TColStd_Array1OfReal & W3, const TColStd_Array1OfReal & W4, const Standard_Boolean isC2 = Standard_False);
 
 };
 
-- 
2.31.1.windows.1

