diff --git a/cmake/UseOpenCASCADE.cmake b/cmake/UseOpenCASCADE.cmake
index 7ddedd7d..965ef302 100644
--- a/cmake/UseOpenCASCADE.cmake
+++ b/cmake/UseOpenCASCADE.cmake
@@ -86,6 +86,7 @@ else(OCE_FOUND)
   if (APPLE)
     find_library (Appkit_LIB NAMES AppKit)
     set_property(TARGET TKOpenGl APPEND PROPERTY IMPORTED_LINK_INTERFACE_LIBRARIES ${Appkit_LIB})
+
   endif(APPLE)
 
 endif(OCE_FOUND)
diff --git a/examples/c_demo/CMakeLists.txt b/examples/c_demo/CMakeLists.txt
index a9dc0f92..d0294d4d 100644
--- a/examples/c_demo/CMakeLists.txt
+++ b/examples/c_demo/CMakeLists.txt
@@ -9,7 +9,7 @@ if (NOT TARGET tigl3)
 endif()
 
 add_executable(c_demo c_demo.c)
-target_link_libraries(c_demo tigl3 ${OpenCASCADE_LIBRARIES})
+target_link_libraries(c_demo tigl3)
 
 install(FILES 
     c_demo.c
diff --git a/src/common/tiglcommonfunctions.cpp b/src/common/tiglcommonfunctions.cpp
index d6e9917c..3492efb7 100644
--- a/src/common/tiglcommonfunctions.cpp
+++ b/src/common/tiglcommonfunctions.cpp
@@ -379,6 +379,8 @@ boost::optional<UVResult> GetFaceAndUV(TopoDS_Shape const& shape,
             }
         }
     }
+
+    return res;
 }
 
 TopoDS_Face TrimFace(TopoDS_Face const& face,
