Index: opencascade-7.6.2/adm/cmake/occt_toolkit.cmake
===================================================================
--- opencascade-7.6.2.orig/adm/cmake/occt_toolkit.cmake
+++ opencascade-7.6.2/adm/cmake/occt_toolkit.cmake
@@ -400,7 +400,7 @@
   if(IS_VTK_9XX)
     string (REGEX REPLACE "vtk" "VTK::" USED_TOOLKITS_BY_CURRENT_PROJECT "${USED_TOOLKITS_BY_CURRENT_PROJECT}")
   endif()
-  target_link_libraries (${PROJECT_NAME} ${USED_TOOLKITS_BY_CURRENT_PROJECT} ${USED_EXTERNAL_LIBS_BY_CURRENT_PROJECT})
+  target_link_libraries (${PROJECT_NAME} PUBLIC ${USED_TOOLKITS_BY_CURRENT_PROJECT} PRIVATE ${USED_EXTERNAL_LIBS_BY_CURRENT_PROJECT})
 endif()
 
 if (USE_QT)
